macros/timeHistoryAnalyzer.py

```diff
@@ -0,0 +1,158 @@
+    import numpy as np
+    import os
+    import os.path
+    import sys
+
+    from optparse import OptionParser, OptionGroup
+    parser = OptionParser()
+    parser.add_option("-i", "--inputDir", type=str, dest="inputDir",
+                      help="Input directory (=output directory of plotTimeSeries.py)")
+    parser.add_option("--ranges", type=str, dest="ranges", default="maskReason",
+                      help="Range selection. Possible values: mask, maskReason, zeroInputCap")
+    parser.add_option("--onlyCurrent", dest="onlyCurrent", action="store_true",
+                      help="Only show ranges that extend until the last scan")
+
+    # Configuration of bad scan recovery
+    parser.add_option("--minScanAvgNoise", type=float, dest="minScanAvgNoise", default=0.1,
+                      help="Minimum noise, averaged over the whole detector, for a scan to be considered")
+    parser.add_option("--maxScanMaskedFrac", type=float, dest="maxScanMaskedFrac", default=0.07,
+                      help="Maximum fraction of masked channel, over the whole detector, for a scan to be considered")
+
+    # Configuration of the range finding algo
+    parser.add_option("--numEndScans", type=int, dest="numEndScans", default=5,
+                      help="Number of 'good' scans to end a range")
+    parser.add_option("--minBadScans", type=int, dest="minBadScans", default=4,
+                      help="Minimum number of 'bad' scans to keep a range")
+    # Only for 'zeroInputCap' range finder
+    parser.add_option("--minNoise", type=float, dest="minNoise", default=0.0414,
+                      help="Lower bound on noise for the 'zeroInputCap' range finder")
+    parser.add_option("--maxNoise", type=float, dest="maxNoise", default=0.109,
+                      help="Upper bound on noise for the 'zeroInputCap' range finder")
+
+    (options, args) = parser.parse_args()
+
+    if options.inputDir is None:
+        print("Error: The -i argument is required")
+        sys.exit(os.EX_USAGE)
+
+    if not os.path.isdir(options.inputDir):
+        print("Error: Not a directory: %s" % options.inputDir)
+        sys.exit(os.EX_USAGE)
+
+    from gempython.gemplotting.utils.anahistory import (
+        findRangesMaskReason,
+        findRangesMask,
+        findRangesZeroInputCap,
+        TimeSeriesData)
+
+    findRangesFct = None
+    findRangesKwArgs = {
+        'numEndScans': options.numEndScans,
+        'minBadScans': options.minBadScans,
+    }
+    if options.ranges == "mask":
+        findRangesFct = findRangesMask
+    elif options.ranges == "maskReason":
+        findRangesFct = findRangesMaskReason
+    elif options.ranges == "zeroInputCap":
+        findRangesFct = findRangesZeroInputCap
+        findRangesKwArgs['minNoise'] = options.minNoise
+        findRangesKwArgs['maxNoise'] = options.maxNoise
+    else:
+        print("Error: Invalid argument for --ranges: %s " % options.ranges)
+        sys.exit(os.EX_USAGE)
```

Comment:
How hard would it be to make the default be "all of the above" (lines 56, 58, and 60) and then if the user specifies `--ranges` it's for a specific.  Is this too complex?  

If it is too complex how hard would it be to provide a script which this executes "all of the above" as three separate calls of `timeHistoryAnalyzer.py` similar to [gemPlotterAllChannels.sh](https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/develop/macros/gemPlotterAllChannels.sh)?

Is this a "less than a day change?" If yes let's include it in this PR.  If not let's make a future feature request on the issue page.

Reply:
Maybe the script is best, and it writes the output to a file since having this spew multiple tables may get confusion?  Please comment when you have a chance.

Reply:
Changing the code would take much less than one day, however it would require significant modifications of the examples.

The main problem is usability:

* The user would have to understand which table come from which `--ranges` value. Since the tables can be large, even a header would be easily missed.
* Printing all tables would quickly fill the terminal buffer.
* The user could wonder why there are overlapping ranges, while it's clear if s.he explicitly asks for another definition

I'm open to do the change.

Reply:
Let's not change the python code.  Instead I think a script `makeAllRanges.sh` or similar would be better that writes the output to a text file; e.g.:

```
timeHistoryAnalyzer.py <args for mask> | tee maskOutput.txt
timeHistoryAnalyzer.py <args for maskReason> | tee maskReasonOutput.txt
timeHistoryAnalyzer.py <args for zeroInputCap> | tee zeroInputCapOutput.txt
```

Or similar.

Reply:
> The main problem is usability:

Alternatively, why are you `print`ing tables, rather than just `write`ing them to a file?

Reply:
Writing to a file isn't necessary atm, >xxx is as easy as -o xxx. If needed later, I'd rather produce trees than writing Markdown tables